Feature/new create account

aerych opened the pull request:
This PR updates new user/blog creation, bringing it inline with the updated sign in flow. Essentially we're porting CreateAccountAndBlogViewController to swift. 

To test: 
Be signed out and choose the Create a Site option on the sign in screen to create a new account and blog. 
- Confirm you can create a new wpcom account and blog successfully. 
- Confirm all fields are required.
- Confirm a mistyped email address is corrected by the EmailChecker.
- Confirm you can not reuse an existing email address, username, or blogname. 
- Confirm the form is completely visible on all supported device sizes, and no fields are obscured by the keyboard.
- Confirm you can successfully save newly created credentials in 1Password

@mattmiklic, this one needs a design eye.

Needs review: @koke could I trouble you for another NUX review?

koke commented on WordPress/Classes/Utility/EmailCheckerHelper.h:
```diff
@@ -0,0 +1,11 @@
+ A proxy for the EmailChecker pod. Allows the EmailChecker functionality to be
+ used in .swift files. 
+ */
++ (NSString *) suggestDomainCorrection:(NSString *)email;
```
I have a Swift version of EmailChecker at https://github.com/koke/EmailChecker-ios
Should we just make the switch? 😁

aerych replied:
I would not be opposed :)

koke commented on WordPress/WordPressApi/WordPressComOAuthClient.h:
```diff
@@ -19,6 +19,7 @@ typedef NS_ENUM(NSUInteger, WordPressComOAuthError) {
 @interface WordPressComOAuthClient : AFHTTPRequestOperationManager
 
 + (WordPressComOAuthClient *)client;
++ (WordPressComOAuthClient *)clientForSwift;
```
Not really sure what's going on here, but you might want to wait for #5271

koke commented on WordPress/Classes/Services/OnePasswordFacade.h:
```diff
@@ -24,7 +32,25 @@ typedef void (^OnePasswordFacadeCallback)(NSString *username, NSString *password
+     passwordGenerationOptions:(NSDictionary *)passwordGenerationOptions
+             forViewController:(UIViewController *)viewController
+                        sender:(id)sender
+                    completion:(void (^)(NSDictionary *loginDictionary, NSError * _Nullable error))completion;
```
Since this is a _facade_, I'd argue that these changes expose too much of the internal details. I'd propose a simpler API and handling all the details inside the facade:

```obj-c
- (void)createLoginForURLString:(NSString *)URLString
                       username:(NSString *)username
                       password:(NSString *)password
                         sender:(id)sender
                     completion:(void (^)(NSString *username, NSString *password, NSError * _Nullable error))completion;
```

koke commented:
The design is definitely different 😁. I think the title needs more whitespace. 

![create-account-before-after-es](https://cloud.githubusercontent.com/assets/8739/14986740/6d8694fe-114d-11e6-91cc-e13ec8b4ec7a.png)

koke commented:
This was the same in the old design, but I thing the 1Password button should be in the username field, and have the "reveal password" button in the password field.

koke commented on WordPress/Classes/ViewRelated/NUX/SigninAccountAndBlogViewController.swift:
```diff
@@ -0,0 +1,656 @@
+    }
+
+
+    /// Configures the appearance of the button to request a 2fa code be sent via SMS.
```
Copy&Paste, I presume? 😛

aerych replied:
Guilty :)

koke commented on WordPress/Classes/ViewRelated/NUX/SigninAccountAndBlogViewController.swift:
```diff
@@ -0,0 +1,656 @@
+
+    func textField(textField: UITextField, shouldChangeCharactersInRange range: NSRange, replacementString string: String) -> Bool {
+        // Disallow spaces except for the password field
+        if textField != passwordField && string == " " {
```
I don't think this form will change soon, but when it does, this seems like a good spot to hide a bug. I'd rather be explicit about which fields disallow spaces rather than "anything but the password"

koke commented on WordPress/Classes/ViewRelated/NUX/SigninAccountAndBlogViewController.swift:
```diff
@@ -0,0 +1,656 @@
+
+/// Create a new WordPress.com account and blog.
+///
+@objc class SigninAccountAndBlogViewController : NUXAbstractViewController, SigninKeyboardResponder
```
I know you're using the `Signin` prefix for all the new classes, but this class name is really confusing. Can we just do `SignupViewController`?

koke commented on WordPress/Classes/ViewRelated/NUX/SigninAccountAndBlogViewController.swift:
```diff
@@ -0,0 +1,656 @@
+    override func viewDidAppear(animated: Bool) {
+        super.viewDidAppear(animated)
+
+        registerForKeyboardEvents(#selector(SigninAccountAndBlogViewController.handleKeyboardWillShow(_:)),
```
Not from this PR, but this needs a label for the first argument.

koke commented on WordPress/Classes/ViewRelated/NUX/SigninAccountAndBlogViewController.swift:
```diff
@@ -0,0 +1,656 @@
+    /// Composes and enqueues the operations to validate the site information,
+    /// user create, user authenticaiton, and blog creation.
+    ///
+    func createUserAndSite() {
```
I'd extract the next few methods somewhere else, let's say `SignupService`?

aerych replied:
Hmm... I'm not sure I agree. We'd be creating a new service just to call other services and the additional abstraction will just make things more complicated.

aerych commented:
> This was the same in the old design, but I think the 1Password button should be in the username field, and have the "reveal password" button in the password field

It did seem odd to have it in the password field when the rest of nux has it in the username field.  Let's change it.

mattmiklic commented:
> I think the title needs more whitespace.

Yeah, here's my suggestion for spacing on the title area.

![mockup](https://cloud.githubusercontent.com/assets/1396552/15027824/ab4e55a8-1213-11e6-8664-3cfbab716e99.png)

<img width="1002" alt="screen shot 2016-05-04 at 4 13 34 pm" src="https://cloud.githubusercontent.com/assets/1396552/15027832/aeb4a3c8-1213-11e6-9498-b2ca10bcd026.png">

One other small nit: typo on "fascinating" at the bottom; and it looks like that text may need to be resized to 13pt.

aerych commented:
Thanks for taking a look and for the mock @mattmiklic.  Here's what I have after the changes. I've checked the measurements with a ruler but please confirm I'm not off by a bit.  These are screenshots from the 5s.  The 6s and 6s Plus will have a bit more of a margin between the nav bar and the icon due to how the layout tries to center the textfields vertically.  Also note that when the keyboard is visible the margin between the icon and navbar shrinks so all text fields remain visible. 

![simulator screen shot may 4 2016 4 46 02 pm](https://cloud.githubusercontent.com/assets/1435271/15030552/993c9ff0-1218-11e6-8cc4-54714aa69e32.png)

![simulator screen shot may 4 2016 4 46 07 pm](https://cloud.githubusercontent.com/assets/1435271/15030556/9d6e2b16-1218-11e6-8d1e-ae1348907473.png)

aerych commented:
Ready for round two @koke

koke commented on WordPress/WordPressApi/WordPressComOAuthClient.m:
```diff
@@ -0,0 +1,148 @@
+static NSString * const WordPressComOAuthBaseUrl = @"https://public-api.wordpress.com/oauth2";
+static NSString * const WordPressComOAuthRedirectUrl = @"https://wordpress.com/";
+
+@implementation WordPressComOAuthClient
```
This was migrated to Swift in #5271. Bad merge?

aerych replied:
Ick. Yah, probably a bad merge. Will fix.

koke commented:
I think the previous version would auto-fill the url when you typed a username.

Also, I tried creating an account and hit an assertion:

```
2016-05-06 12:41:40:896 WordPress[87922:16428447] Failed finishing account creation. The blogOptions dictionary was missing expected data.
fatal error: : file /Users/koke/automattic/wordpress-ios/WordPress/Classes/ViewRelated/NUX/SignupViewController.swift, line 532
(lldb) po blogOptions
▿ 4 elements
  ▿ [0] : 2 elements
    - .0 : "blogid"
    - .1 : 111047394
  ▿ [1] : 2 elements
    - .0 : "blogname"
    - .1 : kokesignup3
  ▿ [2] : 2 elements
    - .0 : "url"
    - .1 : https://kokesignup3.wordpress.com/
  ▿ [3] : 2 elements
    - .0 : "xmlrpc"
    - .1 : https://kokesignup3.wordpress.com/xmlrpc.php
(lldb) po blogOptions["blogid"] as? NSNumber
nil
(lldb) po blogOptions["blogid"]!.dynamicType
NSTaggedPointerString
```

aerych commented:
> I think the previous version would auto-fill the url when you typed a username

Good catch.  I've added this back in. 

> Also, I tried creating an account and hit an assertion

My bad.  I've fixed the cause. 

Ready for another round

mattmiklic commented:
> I've checked the measurements with a ruler but please confirm I'm not off by a bit

That's looking good now; if you could add maybe 2px more margin between the title text and the Username field, I think it would be a little more comfortable. 👍 here though.

aerych commented:
> if you could add maybe 2px more margin between the title text and the Email Address field,

Will do!

koke commented:
> We'd be creating a new service just to call other services and the additional abstraction will just make things more complicated

I suggested a new service, but this could also live inside `AccountService`. We're currently parsing JSON inside a view controller, and handling a lot of logic not related to presentation.

And speaking of all that logic... we should unit test it :)

aerych commented:
@koke I ended up moving most of the logic into a service after all. One last look?

koke commented on WordPress/Classes/Services/SignupService.swift:
```diff
@@ -0,0 +1,347 @@
+    /// An internal struct for conveniently sharing params between the different 
+    /// sign up steps.
+    ///
+    struct SigninParams {
```
`SignupParams` maybe?

aerych replied:
Doh!  Yes, that's what it should be.

koke commented on WordPress/Classes/Services/SignupService.swift:
```diff
@@ -0,0 +1,347 @@
+        var domain : String {
+            return "SignupError"
+        }
+    }
```
If you want to simplify this a bit, Foundation does some magic for you:

```swift
enum SignupError : ErrorType {
    case InvalidResponse
    case MissingRESTAPI
    case MissingDefaultWPComAccount
}

let error = SignupError.MissingDefaultWPComAccount as NSError
print(error)
// Error Domain=SignupError Code=2 "(null)"
```

aerych replied:
Ooo... shiny!

koke commented on WordPress/Classes/Services/SignupService.swift:
```diff
@@ -0,0 +1,347 @@
+                                            failure: SignupFailureBlock) {
+
+        let params = SigninParams(url: url, title: blogTitle, email: emailAddress, username: username, password: password)
+        validateWPComBlogWithParams(params, status: status, success: success, failure: failure)
```
My main concern about the new service is that you are calling a `validate` method that "Validates that the blog can be created and is not already taken", but it will actually create the account and the site.

I think the request chaining should be done in this method and not in the individual requests. It'll look a bit terrible, but it's better than having methods doing more things than what they promise.

aerych commented:
@koke changes made

koke commented on WordPress/Classes/Services/SignupService.swift:
```diff
@@ -54,8 +54,36 @@ public class SignupService : LocalCoreDataService
+
+        // Create call back blocks for the various methods we'll call to create the user account and blog. 
+        // Each success block calls the next step in the process. 
+        // NOTE: The steps below are constructed in reverse order.
```
That was a bit weird at first, but I think it's better than nesting them. Sometimes I dream about the `Future` though 😉

koke commented:
:shipit:

aerych commented:
Thank you Koke!!
